Backup lock

gbrova opened the pull request:
Beginning of work on #852 ; enables sweep to run safely alongside backups

Users of AtlasDB need to coordinate sweep and backup jobs, because running a sweep during a backup can corrupt that backup. This PR adds coordination via a KVS-persisted deletion lock. 

This PR adds a KVS-persisted deletion lock, which sweep and scrub acquire before running. The PR also adds CLIs to manipulate these locks, which can be used to acquire the deletion lock in preparation for a backup. The locks can be exclusive or non-exclusive (similar to read-write locks) - sweep and scrub take a non-exclusive deletion lock (so that sweep and scrub can run simultaneously), while the backup CLI takes an exclusive lock. 

Still need: 

- [ ] Docs and release notes
- [ ] Consider protecting truncates and schema mutations.
- [ ] Support for suppressing the sweep timestamp (to enable partial sweep/scrub during a backup).  Will do this as part of a separate PR.

schlosna commented on atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/BackupLockCommand.java:
```diff
@@ -0,0 +1,68 @@
+    private void blockOnConsoleInput(Console console) {
+        log.info("Successfully acquired the lock. "
+                + "When your backup is complete, press ENTER to release it");
+        console.readLine();
```
curious what happens if one were to run this not inside `screen` or `tmux` and their terminal dies (say `ssh` disconnects), what happens. Is there a recovery mode?

gbrova replied:
`PersistentLockCommand` is a more generic and versatile CLI, we can use that to recover. I've added this second CLI just for ease of use, but we can remove it if there are too many caveats.

schlosna commented on atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/BackupLockCommand.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private void blockOnConsoleInput(Console console) {
+        log.info("Successfully acquired the lock. "
```
Should probably include lock ID once acquired, and probably echo output to console as well.

schlosna commented on atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/PersistentLockCommand.java:
```diff
@@ -0,0 +1,109 @@
+        KeyValueService keyValueService = services.getKeyValueService();
+        PersistentLock persistentLock = new PersistentLock(keyValueService);
+
+        System.out.println("keyValueService = " + keyValueService);
```
I have a feeling this will be a turducken KVS that is wrapped with several layers of delegates & proxies -- is this going to be meaningful to user? What does this currently output?

gbrova replied:
Removed - this was left over from debugging and is totally not useful for users.

schlosna commented on atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/PersistentLockCommand.java:
```diff
@@ -0,0 +1,109 @@
+        try {
+            LockEntry acquiredLock = persistentLock.acquireLock(
+                    PersistentLockName.of(lockName), "PersistentLock CLI command");
+            log.info("Successfully acquired persistent lock " + acquiredLock);
```
nit: slf4j pattern: `log.info("Successfully acquired persistent lock {}",acquiredLock);`

gbrova replied:
Thanks, done.

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    public void releaseLock(LockEntry lock) {
+        log.debug("Releasing persistent lock " + lock);
```
nit: slf4j format

gbrova replied:
Done.

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    private void insertLockEntry(LockEntry lockEntry) {
+        log.debug("Attempting to acquire persistent lock " + lockEntry);
```
nit: slf4j format

gbrova replied:
Done

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java:
```diff
@@ -0,0 +1,161 @@
+
+    private LockEntry verifyLockWasAcquired(LockEntry desiredLock) throws PersistentLockIsTakenException {
+        List<LockEntry> relevantLocks = allRelevantLockEntries(desiredLock.lockName());
+        Preconditions.checkState(relevantLocks.contains(desiredLock), "Lock was not properly inserted");
```
should `relevantLocks` be a set for cheap contains?

gbrova replied:
Done

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java:
```diff
@@ -0,0 +1,161 @@
+        return verifyLockDoesNotConflict(desiredLock, conflictingLocks);
+    }
+
+    private List<LockEntry> retainExclusiveLocks(List<LockEntry> conflictingLocks) {
```
should these be sets in & out?

clockfort replied:
conflicting LockEntrys might hashCode / equals the same?
or maybe we care about ordering here for who wins a conflict

gbrova replied:
Changed to Set throughout

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java:
```diff
@@ -0,0 +1,161 @@
+                .collect(Collectors.toList());
+    }
+
+    private List<LockEntry> removeSingleLock(LockEntry desiredLock, List<LockEntry> relevantLocks) {
```
should these be sets in & out?

gbrova replied:
Done

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java:
```diff
@@ -0,0 +1,161 @@
+            List<LockEntry> conflictingLocks) throws PersistentLockIsTakenException {
+
+        if (conflictingLocks.isEmpty()) {
+            log.debug("Acquired persistent lock " + desiredLock);
```
nit: slf4j format

gbrova replied:
Done

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java:
```diff
@@ -0,0 +1,161 @@
+            log.debug("Acquired persistent lock " + desiredLock);
+            return desiredLock;
+        } else {
+            log.info("Failed to acquire persistent lock " + desiredLock);
```
nit: slf4j format

gbrova replied:
Done.

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java:
```diff
@@ -0,0 +1,161 @@
+        keyValueService.put(AtlasDbConstants.PERSISTED_LOCKS_TABLE, lockEntry.insertionMap(), LOCKS_TIMESTAMP);
+    }
+
+    private List<LockEntry> allRelevantLockEntries(PersistentLockName lock) {
```
return set?

gbrova replied:
Done

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java:
```diff
@@ -0,0 +1,161 @@
+                .collect(Collectors.toList());
+    }
+
+    public List<LockEntry> allLockEntries() {
```
return set?

gbrova replied:
Done

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    public List<LockEntry> allLockEntries() {
+        List<RowResult<Value>> allLockRows = ImmutableList.copyOf(keyValueService.getRange(
```
why copy all in memory first rather than using `BatchingVisitable` to to transform and collection to set?

gbrova replied:
I envision this table containing very few entries at any time, and I find this to be more readable.

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLockIsTakenException.java:
```diff
@@ -0,0 +1,32 @@
+    private final List<LockEntry> conflictingLocks;
+
+    public PersistentLockIsTakenException(List<LockEntry> conflictingLocks) {
+        this.conflictingLocks = conflictingLocks;
```
Should probably just push the message construction into constructor as:

```
public PersistentLockIsTakenException(List<LockEntry> conflictingLocks) {
    super("Could not acquire the requested lock because it conflicts with the " +
          "following other lock(s): " + conflictingLocks);
}
```

Then we don't need to keep the conflictingLocks around (which if you were going to should be a snapshot `ImmutableSet.copyOf(conflictingLocks)` in case the version passed in isn't serializable or mutable).

gbrova replied:
Done

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLockIsTakenException.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Override
+    public String getMessage() {
+        return "Could not acquire the requested lock because it conflicts with the following other lock(s): "
```
see above, should be able to get rid of this

gbrova replied:
Done

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java:
```diff
@@ -116,6 +122,17 @@ public SweepResults run(TableReference tableRef, int batchSize, @Nullable byte[]
         }
 
+        try {
+            String reason = "Sweep for " + tableRef;
```
inline this below so lambda doesn't need to capture both tableRef and reason

gbrova replied:
Done

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java:
```diff
@@ -116,6 +122,17 @@ public SweepResults run(TableReference tableRef, int batchSize, @Nullable byte[]
             return SweepResults.createEmptySweepResult(0L);
         }
 
+        try {
```
might want to extract the try/catch to separate method `runWithDeleteLock` as `run` is getting long

gbrova replied:
Done.  Extracting `runSweepInternal` also made `run` much shorter.

gbrova commented:
@schlosna any other comments? I'll be sure one of @tpetracca or @ashrayjain also approve it before merging.

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    public static LockEntry fromRowResult(RowResult<com.palantir.atlasdb.keyvalue.api.Value> rowResult) {
+        String rowName = new String(rowResult.getRowName(), StandardCharsets.UTF_8);
```
nit: could use `asString(rowResult.getRowName())`

gsheasby commented:
Reviewed 12 of 20 files at r1, 2 of 4 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: 16 of 20 files reviewed at latest revision, 27 unresolved discussions.

---

*[atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/BackupLockCommand.java, line 1 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToM_n53NLwvr4jTclw:-KToM_n6vMPDXg0E2Isx:b-6lw0gv) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/BackupLockCommand.java#L1)):*
> ```Java
> /**
> ```

Ignoring CLIs for now as per chat discussion

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/DeletionLock.java, line 22 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToNiKE6ZyCInZM9xQ7:-KToNiKE6ZyCInZM9xQ8:bmimfc5) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/DeletionLock.java#L22)):*
> ```Java
> import com.palantir.atlasdb.keyvalue.api.KeyValueService;
> 
> public class DeletionLock {
> ```

I wondered for a moment why we have this class if it just forwards to `PersistentLock`. I guess the motivation is that with this class, we can change how the deletion lock works in future without an API break.

Examples of how we could change it: maybe use something other than a `PersistentLock` for some reason, or do extra checking, or something allowing partial sweep.

The last one we probably want to do, according to the PR blurb.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/DeletionLock.java, line 34 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToOCjlcyBtpsK8szzg:-KToOCjlcyBtpsK8szzh:bz0omw6) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/DeletionLock.java#L34)):*
> ```Java
>     }
> 
>     public <T> T  runWithLockNonExclusively(Supplier<T> supplier, String reason) throws PersistentLockIsTakenException {
> ```

What does this mean? If I'm a dev using the deletion lock, when would I want to run non-exclusively?

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java, line 68 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToPRj3wQC6W1gUKtj_:-KToPRj3wQC6W1gUKtja:bxlkh97) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java#L68)):*
> ```Java
>     }
> 
>     public static String valueOfColumnInRow(
> ```

This looks like it should be handled by some Cassandra utils class. Or possibly an object whose responsibility is to get stuff out of `RowResult`s, as per the discussion in @jboreiko's conference session yesterday (prefer tested objects to static library methods).

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java, line 76 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToQWXOwQRwfjTBg7sy:-KToQWXOwQRwfjTBg7sz:b-f6unre) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java#L76)):*
> ```Java
>                 .get();
> 
>         byte[] contents = rowResult.getColumns().get(columnCell.getColumnName()).getContents();
> ```

observation, beyond the scope of this PR (OBSOP?): ugh. The law of demeter feels violated. Possible future target for refactoring.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java, line 114 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToR2FL-xNJxPGJpZfw:-KToR2FMSyLWD-iyuRni:bbm80am) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java#L114)):*
> ```Java
>     }
> 
>     private static String asString(byte[] value) {
> ```

I know I just railed against them, but isn't this a utility method somewhere?

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 109 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KSvLyS-j5FjQUexctDn-r1-109:-KToWJ-AWhaiWEey-Ya2:bjtzipd) ([raw file](https://github.com/palantir/atlasdb/blob/46eab6960c7878b41f8b34a4962fdc5c7883df90/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L109)):*
<details><summary><i>Previously, gbrova (George Brova) wrote…</i></summary>
> Changed to Set throughout
</details>
LGTM. Leaving for @clockfort to ack.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 147 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KSvLyS-j5FjQUexctDn-r1-147:-KToX-hyt0ShJ3CNwdj9:b-bdfkiv) ([raw file](https://github.com/palantir/atlasdb/blob/46eab6960c7878b41f8b34a4962fdc5c7883df90/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L147)):*
<details><summary><i>Previously, gbrova (George Brova) wrote…</i></summary>
> I envision this table containing very few entries at any time, and I find this to be more readable. 
</details>
There should be at most one entry per lock type, right? And currently we have one lock type?

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 45 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToRdO7xhz6UBMcx9Ig:-KToRdO8VUDAYjJumKM9:bewztm3) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L45)):*
> ```Java
>         this.keyValueService = keyValueService;
> 
>         createPersistedLocksTableIfNotExists();
> ```

Code smell: Behaviour In Constructor

It feels quite ugly to have this behaviour in a constructor, and might make testing harder (I haven't got to the tests yet, so this might be a moot point).

Consider having a static factory method `create()` and a private or VisibleForTesting constructor, as dictated by the tests.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 82 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToSeGz__2j26cFpyL-:-KToSeGz__2j26cFpyL0:b-uia2y4) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L82)):*
> ```Java
>     }
> 
>     public LockEntry acquireLock(
> ```

Does this need to be public?

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 85 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToTADk_aS4qZp1DeUP:-KToTADk_aS4qZp1DeUQ:b-cfig4q) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L85)):*
> ```Java
>             PersistentLockName lockName, String reason, boolean exclusive) throws PersistentLockIsTakenException {
>         long thisId = ThreadLocalRandom.current().nextLong(Long.MAX_VALUE);
>         LockEntry lockEntry = LockEntry.of(lockName, thisId, reason, exclusive);
> ```

very mild nit: These two lines could actually go in a method called `generateUniqueLockEntry`, especially if the logic is repeated elsewhere.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 101 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToTi2njrFbKOujtwkH:-KToTi2njrFbKOujtwkI:bjricwr) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L101)):*
> ```Java
>         Preconditions.checkState(relevantLocks.contains(desiredLock), "Lock was not properly inserted");
> 
>         Set<LockEntry> conflictingLocks = removeSingleLock(desiredLock, relevantLocks);
> ```

Could just use `Sets.difference`. The method name confused me for a moment.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 101 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToVr2s5N4WmM6P4I53:-KToVr2tXZVb4KFm1F9E:biz3s9m) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L101)):*
> ```Java
>         Preconditions.checkState(relevantLocks.contains(desiredLock), "Lock was not properly inserted");
> 
>         Set<LockEntry> conflictingLocks = removeSingleLock(desiredLock, relevantLocks);
> ```

Could you push this logic down into `verifyLockDoesNotConflict`, please?

Perhaps that method would become:
```
Set<LockEntry> conflictingLocks = getConflictingLocks(...);

if (conflictingLocks.isEmpty()) {
   ....
```

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 135 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToWQkH_U7cDiB4gQbZ:-KToWQkH_U7cDiB4gQb_:b9nrd2z) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L135)):*
> ```Java
>     }
> 
>     private void insertLockEntry(LockEntry lockEntry) {
> ```

Nit: looks like this has become a long way removed from its use. Consider moving it up the file.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 146 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToWoIhH7_kIyufS3MA:-KToWoIhH7_kIyufS3MB:b-uia2y4) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L146)):*
> ```Java
>     }
> 
>     public Set<LockEntry> allLockEntries() {
> ```

Does this need to be public?

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 156 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToX8o2du5COOpb4jmX:-KToX8o2du5COOpb4jmY:b-sgdotl) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L156)):*
> ```Java
> 
>     private void createPersistedLocksTableIfNotExists() {
>         if (!keyValueService.getAllTableNames().contains(AtlasDbConstants.PERSISTED_LOCKS_TABLE)) {
> ```

The key value service might handle this case internally now. I think that's the case for Cassandra KVS, anyway.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java, line 125 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KSvLyRzZ2fu1NoPwmJF-r1-125:-KToZHVwI-zAYnTtVGG7:b-8fx2u7) ([raw file](https://github.com/palantir/atlasdb/blob/46eab6960c7878b41f8b34a4962fdc5c7883df90/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java#L125)):*
<details><summary><i>Previously, gbrova (George Brova) wrote…</i></summary>
> Done.  Extracting `runSweepInternal` also made `run` much shorter. 
</details>
:+1: 

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java, line 126 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KSvLyRzZ2fu1NoPwmJF-r1-126:-KToZOUeqH8hcX302YMJ:bb6mt9b) ([raw file](https://github.com/palantir/atlasdb/blob/46eab6960c7878b41f8b34a4962fdc5c7883df90/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunnerImpl.java#L126)):*
<details><summary><i>Previously, gbrova (George Brova) wrote…</i></summary>
> Done
</details>
:+1:

---

*[atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLockNameShould.java, line 20 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToZViHGLbWT3uQBXeW:-KToZViHGLbWT3uQBXeX:b-hdcpxo) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLockNameShould.java#L20)):*
> ```Java
> import org.junit.Test;
> 
> public class PersistentLockNameShould {
> ```

It's a trivial test, but how about `allowNameWithoutDelimiter`?

---

*[atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java, line 105 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KTo_mLgOAD5HIbmB5Ru:-KTo_mLgOAD5HIbmB5Rv:bspaudb) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java#L105)):*
> ```Java
> 
>     @Test
>     public void runActionWhileLockIsHeld() throws PersistentLockIsTakenException {
> ```

I believe this test would still pass even if we didn't run the action. Might I suggest having a variable declared before `runWithExclusiveLock`, incremented in the action, and checked after `runWithExclusiveLock`?

---

*[atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java, line 127 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToa21Ur6F-xfyZpyot:-KToa21Ur6F-xfyZpyou:bg5gjnv) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java#L127)):*
> ```Java
>         PersistentLockName deletionLock = PersistentLockName.of("deletionLock");
>         LockEntry existingLock = LockEntry.of(PersistentLockName.of("deletionLock"), 4321, REASON);
>         keyValueService.put(AtlasDbConstants.PERSISTED_LOCKS_TABLE, existingLock.insertionMap(), 0);
> ```

Did you consider using `when` here, like in an above test?

---

*[atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java, line 174 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToaLzMeB9IfkdukjVV:-KToaLzMeB9IfkdukjVW:b-7vrk9w) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java#L174)):*
> ```Java
> 
>     @Test
>     public void allowNonExclusiveLocksToBeTakenOutConcurrently() throws PersistentLockIsTakenException {
> ```

Could also have a test to prove that you can take out `deletionLock` when `anotherLock` is held - even if they're exclusive.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1011)*
<!-- Sent from Reviewable.io -->

gsheasby commented:
Review status: 16 of 20 files reviewed at latest revision, 27 unresolved discussions.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java, line 31 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToagG52P0p4UhGB8px:-KToagG6iMFvKYQ-_jZS:b-v1jdy8) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java#L31)):*
> ```Java
> 
> @Value.Immutable
> public abstract class LockEntry {
> ```

There do not appear to be tests for this class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1011)*
<!-- Sent from Reviewable.io -->

gbrova commented:
*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/DeletionLock.java, line 22 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToNiKE6ZyCInZM9xQ7:-KTpj0__Rgc3hBsy8WAt:b-9ajfv3) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/DeletionLock.java#L22)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> I wondered for a moment why we have this class if it just forwards to `PersistentLock`. I guess the motivation is that with this class, we can change how the deletion lock works in future without an API break.
> 
> Examples of how we could change it: maybe use something other than a `PersistentLock` for some reason, or do extra checking, or something allowing partial sweep.
> 
> The last one we probably want to do, according to the PR blurb.
</details>
Yes, that's one thing we might do in the future. More generally, I wanted a small wrapper around Lock to expose the simplest API possible.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1011)*
<!-- Sent from Reviewable.io -->

gbrova commented:
Review status: 11 of 20 files reviewed at latest revision, 38 unresolved discussions.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/DeletionLock.java, line 34 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToOCjlcyBtpsK8szzg:-KTpjpDyrFyakiUGS0PQ:brux3iu) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/DeletionLock.java#L34)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> What does this mean? If I'm a dev using the deletion lock, when would I want to run non-exclusively?
</details>
The concept is similar to a read-write lock.  In particular, you might want to run multiple sweep/scrub jobs concurrently (and possibly also deletions / schema mutations, if we decide to protect those in a future PR).

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java, line 61 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KSvLyS-j5FjQUexctDm-r6-61:-KTpucUI46PjzSV-urdz:bb6un1j) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java#L61)):*
<details><summary><i>Previously, schlosna (David Schlosnagle) wrote…</i></summary>
> nit: could use `asString(rowResult.getRowName())`
</details>
Done

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 147 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KSvLyS-j5FjQUexctDn-r1-147:-KTponojtyZeLfSKFtz9:bdsa040) ([raw file](https://github.com/palantir/atlasdb/blob/46eab6960c7878b41f8b34a4962fdc5c7883df90/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L147)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> There should be at most one entry per lock type, right? And currently we have one lock type?
</details>
Possibly multiple if they are non-exclusive; even then, the number should be small. 

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 45 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToRdO7xhz6UBMcx9Ig:-KTptnHGgGMwB299uh8t:bb6un1j) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L45)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Code smell: Behaviour In Constructor
> 
> It feels quite ugly to have this behaviour in a constructor, and might make testing harder (I haven't got to the tests yet, so this might be a moot point).
> 
> Consider having a static factory method `create()` and a private or VisibleForTesting constructor, as dictated by the tests.
</details>
Done

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 82 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToSeGz__2j26cFpyL-:-KTpljrwB_yPkO2v6xqs:b-x7257e) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L82)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Does this need to be public?
</details>
It doesn't currently, but I don't see any advantage to hiding it given that we are exposing `releaseLock`. For example, we might want the CLI to support acquiring non-exclusive locks. 

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 85 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToTADk_aS4qZp1DeUP:-KTpmjQ7fJI3TIQE6ZwT:b-896fix) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L85)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> very mild nit: These two lines could actually go in a method called `generateUniqueLockEntry`, especially if the logic is repeated elsewhere.
</details>
Done.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 101 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToTi2njrFbKOujtwkH:-KTpnErpPRTQLQjR01G5:bb6un1j) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L101)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Could just use `Sets.difference`. The method name confused me for a moment.
</details>
Done

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 101 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToVr2s5N4WmM6P4I53:-KTpo4ndQZHbSIsjYnHF:bb6un1j) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L101)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Could you push this logic down into `verifyLockDoesNotConflict`, please?
> 
> Perhaps that method would become:
> ```
> Set<LockEntry> conflictingLocks = getConflictingLocks(...);
> 
> if (conflictingLocks.isEmpty()) {
>    ....
> ```
</details>
Done

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 135 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToWQkH_U7cDiB4gQbZ:-KTpoZtvI1zpdYqMChCh:bb6un1j) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L135)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Nit: looks like this has become a long way removed from its use. Consider moving it up the file.
</details>
Done

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 146 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToWoIhH7_kIyufS3MA:-KTpoeRLxEiybaAUHwwi:b-7qimb3) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L146)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Does this need to be public?
</details>
Yes, I want to expose it in the CLI

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 156 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToX8o2du5COOpb4jmX:-KTppOmyWbQtOPGAhvRT:b-yoej1j) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L156)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> The key value service might handle this case internally now. I think that's the case for Cassandra KVS, anyway.
</details>
You're right, the KVS API even documents this. Removed check. 

---

*[atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLockNameShould.java, line 20 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToZViHGLbWT3uQBXeW:-KTppkQ4IpSz2L4rQNSZ:b-1eh3wi) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLockNameShould.java#L20)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> It's a trivial test, but how about `allowNameWithoutDelimiter`?
</details>
Sure, done.

---

*[atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java, line 105 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KTo_mLgOAD5HIbmB5Ru:-KTpq-B4gPLd7UsOoz65:b-o7vuhc) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java#L105)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> I believe this test would still pass even if we didn't run the action. Might I suggest having a variable declared before `runWithExclusiveLock`, incremented in the action, and checked after `runWithExclusiveLock`?
</details>
Good catch! Done.

---

*[atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java, line 127 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToa21Ur6F-xfyZpyot:-KTpqw6f7WJQIRV4drc3:b67p56j) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java#L127)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Did you consider using `when` here, like in an above test?
</details>
I considered it, but it's tedious and fragile to write out all the `when` and `thenReturn` for both the puts and the gets. 

---

*[atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java, line 174 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToaLzMeB9IfkdukjVV:-KTpsCF1far85Hig_FAT:b-sjgzx2) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java#L174)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Could also have a test to prove that you can take out `deletionLock` when `anotherLock` is held - even if they're exclusive.
</details>
You should _not_ be able to take out an exclusive lock if another (exclusive or nonexclusive) lock exists. 

I've added `forbidExclusiveLockIfAnotherExclusiveLockIsTaken` for symmetry. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1011)*
<!-- Sent from Reviewable.io -->

gbrova commented:
@gsheasby Thanks for the review; I addressed most of your comments, will take care of LockEntry tomorrow.

gsheasby commented:
Reviewed 5 of 7 files at r7, 3 of 3 files at r8.
Review status: 17 of 21 files reviewed at latest revision, 12 unresolved discussions.

---

*[atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java, line 174 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToaLzMeB9IfkdukjVV:-KTsCrs1klShekS_9YFb:b-lsgg5q) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java#L174)):*
<details><summary><i>Previously, gbrova (George Brova) wrote…</i></summary>
> You should _not_ be able to take out an exclusive lock if another (exclusive or nonexclusive) lock exists. 
> 
> I've added `forbidExclusiveLockIfAnotherExclusiveLockIsTaken` for symmetry. 
</details>
I meant, what if they have a different name, i.e. what happens with the below?
```
        persistentLock.acquireLock(PersistentLockName.of("anotherLock"), REASON, true);
        persistentLock.acquireLock(PersistentLockName.of("deletionLock"), REASON, true);
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1011)*
<!-- Sent from Reviewable.io -->

gbrova commented:
Review status: 15 of 21 files reviewed at latest revision, 25 unresolved discussions.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java, line 31 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToagG52P0p4UhGB8px:-KTt5j3WSA015N0Z4JrV:b-8b11j9) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java#L31)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> There do not appear to be tests for this class.
</details>
Added

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java, line 68 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToPRj3wQC6W1gUKtj_:-KTsPHRUZOaTSsJvUs8i:b-y7gr5u) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java#L68)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> This looks like it should be handled by some Cassandra utils class. Or possibly an object whose responsibility is to get stuff out of `RowResult`s, as per the discussion in @jboreiko's conference session yesterday (prefer tested objects to static library methods).
</details>
We can actually remove a lot of this logic, I think it's cleaner/simpler now.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java, line 76 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToQWXOwQRwfjTBg7sy:-KTt5oGIINLFenibtVja:bqt39dn) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java#L76)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> observation, beyond the scope of this PR (OBSOP?): ugh. The law of demeter feels violated. Possible future target for refactoring.
</details>
Agreed, it might be useful to refactor or move things into RowResult (for example) at some point in the future. 

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java, line 114 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToR2FL-xNJxPGJpZfw:-KTt6BzMMz6Tq_cLTVCD:b-62ygmo) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockEntry.java#L114)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> I know I just railed against them, but isn't this a utility method somewhere?
</details>
It's in the Cassandra code somewhere, but I don't want to use that because this isn't Cassandra-specific. TBH, it's simple enough that I prefer having it here instead of a utility, though maybe eventually we can 

---

*[atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java, line 174 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToaLzMeB9IfkdukjVV:-KTsLwH_GdlOG4rj-fsB:b2atir0) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/persistentlock/PersistentLocksShould.java#L174)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> I meant, what if they have a different name, i.e. what happens with the below?
> ```
>         persistentLock.acquireLock(PersistentLockName.of("anotherLock"), REASON, true);
>         persistentLock.acquireLock(PersistentLockName.of("deletionLock"), REASON, true);
> ```
</details>
Ah, I misunderstood. Added `allowDifferentLocksToBeTakenOutConcurrently`.

---

*[atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/BackupLockCommand.java, line 63 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KSvLyRzZ2fu1NoPwmJI-r1-63:-KTtTM6gkgmez0Tu-k5M:b-glkx8u) ([raw file](https://github.com/palantir/atlasdb/blob/46eab6960c7878b41f8b34a4962fdc5c7883df90/atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/BackupLockCommand.java#L63)):*
<details><summary><i>Previously, schlosna (David Schlosnagle) wrote…</i></summary>
> Should probably include lock ID once acquired, and probably echo output to console as well.
</details>
Will delete `PersistentLockCommand` entirely in preference to the more generic `PersistentLockCommand`. 

---

*[atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/BackupLockCommand.java, line 1 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KToM_n53NLwvr4jTclw:-KTtT9sN-87lDQaO5KM2:b1xwgjf) ([raw file](https://github.com/palantir/atlasdb/blob/8973fa41848252de34f84e91d5098a8bc9844234/atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/BackupLockCommand.java#L1)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Ignoring CLIs for now as per chat discussion
</details>
You can look at the CLI too now (I initially had two but I think keeping only one makes more sense since, at least initially, only one internal script will call this)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1011)*
<!-- Sent from Reviewable.io -->

gsheasby commented:
Reviewed 2 of 7 files at r7, 5 of 5 files at r9.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/PersistentLockCommand.java, line 76 at r9](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KTtemUPQJPBrl5U-5re:-KTtemUPQJPBrl5U-5rf:b-fnn1n5) ([raw file](https://github.com/palantir/atlasdb/blob/ca6fdc280aac33424c7a4c6dfd0d2bbd872ea167/atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/PersistentLockCommand.java#L76)):*
> ```Java
>         } else {
>             log.error("See the help page for instructions on how to use this CLI");
>             return 1;
> ```

 You should print the usage in both of the log.error cases.

---

*[atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/PersistentLockCommand.java, line 103 at r9](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KTtf58zOJFQg-ftOpRC:-KTtf58zOJFQg-ftOpRD:b-9ny7ml) ([raw file](https://github.com/palantir/atlasdb/blob/ca6fdc280aac33424c7a4c6dfd0d2bbd872ea167/atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/PersistentLockCommand.java#L103)):*
> ```Java
>         log.info("The following persistent locks are currently taken out (total {})", allLockEntries.size());
>         for (LockEntry lockEntry : allLockEntries) {
>             log.info("  {}", lockEntry.toString());
> ```

Nit: the double space is a bit random.

---

*[atlasdb-cli/src/test/java/com/palantir/atlasdb/cli/command/PersistentLockCommandTest.java, line 85 at r9](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KTtfRi7kXu4IUGN_-g4:-KTtfRi7kXu4IUGN_-g5:bhfej4z) ([raw file](https://github.com/palantir/atlasdb/blob/ca6fdc280aac33424c7a4c6dfd0d2bbd872ea167/atlasdb-cli/src/test/java/com/palantir/atlasdb/cli/command/PersistentLockCommandTest.java#L85)):*
> ```Java
>             assertThat(stdout, containsString(LOCK_NAME.name()));
>         }
>     }
> ```

Missing a test for releaseLock.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1011)*
<!-- Sent from Reviewable.io -->

gbrova commented:
Review status: all files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/PersistentLockCommand.java, line 76 at r9](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KTtemUPQJPBrl5U-5re:-KU2iWIOrLTY4gMiGGSF:bb6un1j) ([raw file](https://github.com/palantir/atlasdb/blob/ca6fdc280aac33424c7a4c6dfd0d2bbd872ea167/atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/PersistentLockCommand.java#L76)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
>  You should print the usage in both of the log.error cases.
</details>
Done

---

*[atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/PersistentLockCommand.java, line 103 at r9](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KTtf58zOJFQg-ftOpRC:-KU290P7KsiYDX9_FMcI:b-bul574) ([raw file](https://github.com/palantir/atlasdb/blob/ca6fdc280aac33424c7a4c6dfd0d2bbd872ea167/atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/PersistentLockCommand.java#L103)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Nit: the double space is a bit random.
</details>
Replaced with ` - `. 

---

*[atlasdb-cli/src/test/java/com/palantir/atlasdb/cli/command/PersistentLockCommandTest.java, line 85 at r9](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KTtfRi7kXu4IUGN_-g4:-KU2ZZ6p9T-_lBV44gsR:bqmjwcz) ([raw file](https://github.com/palantir/atlasdb/blob/ca6fdc280aac33424c7a4c6dfd0d2bbd872ea167/atlasdb-cli/src/test/java/com/palantir/atlasdb/cli/command/PersistentLockCommandTest.java#L85)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Missing a test for releaseLock.
</details>
Added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1011)*
<!-- Sent from Reviewable.io -->

gsheasby commented:
Reviewed 3 of 3 files at r10.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1011)*
<!-- Sent from Reviewable.io -->

gsheasby commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-:-KU38qGi1hsgBimTf_0m:bnfp4nl)*
<!-- Sent from Reviewable.io -->

gsheasby commented:
LGTM pending @tpetracca's seal of approval

tpetracca commented on docs/source/cluster_management/backup-restore.rst:
```diff
@@ -61,10 +76,16 @@ Next, we want to clean out any transactions that were committed after our backup
 
      $ ./bin/atlasdb --config <config-file> --config-root <path-to-atlas-block> timestamp fast-forward --file <backup-directory>/fast-forward.timestamp
 
+Finally, the backup lock is persisted to disk and may have been saved as part of the backup.  We can release it with
```
I wouldn't have even think about this. Good catch.

gbrova commented:
Discussed offline with @tpetracca. It could be worthwhile to make this highly available by avoiding / delaying the kvs deletion (which requires quorum all). Considering doing that work as a separate PR though, since this currently only affects sweep (which requires quorum all anyway)

gbrova commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java:
```diff
@@ -0,0 +1,182 @@
+
+    public void releaseLock(LockEntry lock) {
+        log.debug("Releasing persistent lock {}", lock);
+        keyValueService.delete(AtlasDbConstants.PERSISTED_LOCKS_TABLE, lock.deletionMapWithTimestamp(LOCKS_TIMESTAMP));
```
It will not suffer from that issue because each row in this table has a different lockId.

gbrova commented on atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/BackupLockCommand.java:
```diff
@@ -0,0 +1,80 @@
+        KeyValueService keyValueService = services.getKeyValueService();
+        PersistentLock persistentLock = PersistentLock.create(keyValueService);
+
+        if (acquireLock) {
```
Good catch, done.

tpetracca commented:
@gbrova want to start an issue to scope out the things you and I discussed re moving from delete to putting a cleared value and making the deletes happen off thread in a non-critical path way?

tpetracca commented:
Otherwise LGTM

gbrova commented:
Thanks @tpetracca!  Instead of starting an issue, I put up a PR directly (see #1099) because it turns out the change wasn't that hard to do.

clockfort commented:
Review status: 6 of 26 files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java, line 109 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/1011#-KSvLyS-j5FjQUexctDn-r1-109:-KVgIrmsVrw52z33fytx:b-3yplct) ([raw file](https://github.com/palantir/atlasdb/blob/46eab6960c7878b41f8b34a4962fdc5c7883df90/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java#L109)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> LGTM. Leaving for @clockfort to ack.
</details>
I didn't look very far back in this code, but I'm not sure all of the usages of equals/hashCode/sets are correct;

ImmutableLockEntry equals/hashcode is lock name / lock ID / reason / exclusive-boolean all the same.

I would have expected the definition of conflicting locks to be more like "lock names are the same and at least one is in exclusive mode"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1011)*
<!-- Sent from Reviewable.io -->

tpetracca commented:
I don't think conflicting lock and locks equaling each other are intended to be the same thing the way it's written (unless I misunderstood what you were getting at)

gsheasby commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockStore.java:
```diff
@@ -0,0 +1,106 @@
+
+    public void insertLockEntry(LockEntry lockEntry) {
+        log.debug("Attempting to acquire persistent lock {}", lockEntry);
+        keyValueService.put(AtlasDbConstants.PERSISTED_LOCKS_TABLE, lockEntry.insertionMap(), LOCKS_TIMESTAMP);
```
Copying from before:
@jboreiko @tpetracca this seems fishy to me. Inserting the same value twice succeeds. At such a low level, is this a correctness issue?

gsheasby replied:
After discussion offline - this kind of insertion _is_ a correctness issue.
